feat: 8.4 release post (#148)
hamlet-io/docs: blog/2021-12-16-Release-v8.4.md

```bash
hamlet task run-runbook -n management-ssh_ssh-session-runbook
```

```terminal
[*] management-ssh_ssh-session-runbook
[*]   Starts an interactive SSH session with the bastion host using the baseline ssh key

[-] Running contract stage aws_login
      Step: aws_login - Task: set_provider_credentials
[-] Running contract stage ssh_key_decrypt
      Step: ssh_key_decrypt - Task: aws_decrypt_kms_ciphertext
[-] Running contract stage start_ssh_shell
      Step: start_ssh_shell - Task: ssh_run_command

---- SSH Session | host: 1.2.3.4 | user: ec2-user ----

[ec2-user@ip-10-0-245-144 ~]$
```

Each step shows what is being run along with any output that is generated and the cli will pass through interactive sessions as required and Inputs, when required, can be added as key=value arguments on the cli command.

If you'd like to include the runbooks from the list-runbooks command in your existing solutions they are available in the modules:

- [ssh_bastion](https://github.com/hamlet-io/engine-plugin-aws/blob/master/aws/modules/ssh_bastion)
- [runbook_rds_pgdump](https://github.com/hamlet-io/engine-plugin-aws/tree/master/aws/modules/runbook_rds_pgdump)

## Test tooling Updates

> With deployment testing we have made a couple of updates to make testing easier and provide better control over the configuration of tests

Reviewer: ```suggestion
With deployment testing we have made a couple of updates to make testing easier and provide better control over the configuration of tests:-
```

- cfn_nag has been replaced with [checkov](https://www.checkov.io/) which provides security best practice auditing for a wide range of declarative infrastructure templates include CloudFormation and ARM templates. checkov is also python based so we can include it as part of our cli installation

> - When configuring testing tools you can now provide configuration options to the testing tools. This allows you to disable particular checks or enforce the use of a particular framework format when working with checkov

Reviewer: ```suggestion
- When configuring testing tools you can now provide configuration options to the testing tools. This allows you to disable particular checks or enforce the use of a particular framework format when working with checkov.
```

## Stack Management Updates

AWS Stack management now handles failed creations and will remove the existing stack before attempting a new creation. This saves having to go into the console when a deployment fails on the initial creation.

> Azure deployments have been updated to more closely align with the changes to the AWS stack management processes to ensure that a consistent approach is used across the two  providers
> 

Reviewer: ```suggestion
Azure deployments have been updated to more closely align with the changes to the AWS stack management processes to ensure that a consistent approach is used across the two  providers.

```

## Executor Updates

> We've made a range of changes to both the cli and bash backend to make it easier to work with hamlet

Reviewer: ```suggestion
We've made a range of changes to both the cli and bash backend to make it easier to work with hamlet:-
```

> - dos2unix has been removed from bash scripts, if this is required it will need to managed outside of hamlet from now on. Git mostly handles this process anyway

Reviewer: ```suggestion
- dos2unix has been removed from bash scripts. If this is required, it will need to be managed outside of hamlet from now on. Git mostly handles this process anyway
```

- We've updated the cli framework that we use (Click) to its latest major release and as part of that made the autocomplete setup a dedicated command. To enable autocomplete run the following command

    ```bash
    hamlet autocomplete <your shell of choice>
    ```

    Currently supports bash, fish and zsh, after running the command you may need to restart your shell for it to take affect. Note that bash is still used by all hamlet commands in the background, but you can use any shell you like when making calls through the hamlet cli.
- Location directory checks and authentication for the automation scripts have been removed from the bash executor. This means you can run hamlet commands from anywhere as long as the context is provided
- MacOS support has had a couple of updates to ensure that commands in the bash executor align with what is available in the OS
  - Sleep commands have been updated to only specify the time as a number
  - Run id generation which relied on using dd has been replaced with the builtin $RANDOM builtin bash function + base64 encoding
  - Testing of the cli for development no longer uses mknod which is considered a privileged command in MacOS

> - If your are using the URL or containerregsitry image sources you can skip pulling the image by setting the env var `HAMLET_SKIP_IMAGE_PULL=true` when running a deployment. This is handy when you are verifying a deployment

Reviewer: ```suggestion
- If your are using the URL or containerregistry image sources you can skip pulling the image by setting the env var `HAMLET_SKIP_IMAGE_PULL=true` when running a deployment. This is handy when you are verifying a deployment
```

## Component Updates

> This release we added a couple of components and made some minor updates to align naming

Reviewer: ```suggestion
This release we added a couple of components and made some minor updates to align naming:-
```

- outbound mta: The MTA component now supports both inbound and outbound mail handling using the same rule based approach for each direction
- client vpn support: clientvpns define a way for users to access a private network through an authenticated encrypted session from their clients
- LB Conditions: Load balancer conditions have been extended to support a range of different conditions, including HTTP standard details like Headers and query strings along with IP Addresses

> - EFS rename to fileshare: the EFS component has been renamed to fileshare to provide vendor agnostic naming for the component and allow for implementing other engines in the same fileshare functionality

Reviewer: ```suggestion
- EFS rename to fileshare: the EFS component has been renamed to fileshare to provide vendor agnostic naming for the component and allow for implementing other engines in the same fileshare functionality.
```

### AWS

> The AWS plugin has had a range of updates to support different services on our current components

Reviewer: ```suggestion
The AWS plugin has had a range of updates to support different services on our current components:-
```

- wafV2 for AWS: For all WAF configuration options you can now provide a Version attribute to enable WAFv2 over the current classic WAF
- Dynamic Path support for kinesis firehose allows you to control where a record is placed in s3 based on details found during processing of a record

> - ECS Service/Task Now supports specifying the number of GPUs required for a Container
> - MTA inbound and out support has been implemented
> - Filters can be used on topic subscriptions to control the messages sent to a particular subscription

Reviewer: ```suggestion
- ECS Service/Task now supports specifying the number of GPUs required for a Container
- MTA inbound and out support has been implemented
- Filters can be used on topic subscriptions to control the messages sent to a particular subscription
```

- For aws plugin development we have now enabled cfn-lint across all of our test generation to ensure that we comply with cfn-lint
- The directory component now supports AWS AD Connectors which act as proxy services to existing AD DS installations
- EFS now supports creating FSX for SMB file shares

> - When deploying a directory component a route53 resolver is added to route DNS requests from the standard AWS DNS server to the AD servers. This removes the need for creating DHCP chagnes or manually configuring your DNS servers on instances

Reviewer: ```suggestion
- When deploying a directory component a route53 resolver is added to route DNS requests from the standard AWS DNS server to the AD servers. This removes the need for creating DHCP changes or manually configuring your DNS servers on instances
```

- Application LB now supports additional conditions when applying rules
- CFN-HUP support has been added for linux and windows based EC2 instances to re-run cfn-init scripts on CloudFormation Updates

### Azure

> The Azure plguin has had some updates focussing on Networking and directories

Reviewer: ```suggestion
The Azure plugin has had some updates focussing on networking and directories:-
```

### Azure

The Azure plugin has had some updates focussing on networking and directories:-

- Adds support for the directory component using Azure AD Directory Services which creates an AD DS installation based on your Azure AD
- The network component has been refactored significantly to provide more control over Network ACLS and the provisioning of network resources
- Site to Site VPN support has been added to the Gateway component

## Major Fixes

A lot of bugs where taken care of during this release, which you can find in the change logs but the following are notable fixes:

- **s3 versioning without lifecycle management:** When you enable versioning on an S3 component, lifecycle rules are automatically enabled to remove files from the bucket based on the default retention period of the environment. As a result objects won't be visible to applications and appear to have been deleted. Since enabling versioning keeps the files after they are deleted, data will not be lost but files may need to be reverted to existing versions. Please check that your S3 buckets in AWS are working as expected.

> - **orphaned deployments are always first:** A bug around the orphaned deployment processing meant that they were not being found. This has now been resolved and orphaned deployments will now be processed before other deployments to ensure they don't cause issues when running updates.

Reviewer 2: ```suggestion
- **orphaned deployments are always first:** A bug related to orphaned deployment processing meant that orphaned deployments  were not being found. This has now been resolved and orphaned deployments will now be processed before other deployments to ensure they don't cause issues when running updates.
```

## Full Change Log

> To see the full set of changes check out the CHANGELOGS for each of the hamlet services below

Reviewer 2: ```suggestion
To see the full set of changes check out the CHANGELOGS for each of the hamlet services below:
```
